xapian-applications/omega/handler_libextractor.cc

```diff
@@ -0,0 +1,120 @@
+/** @file handler_libextractor.cc
+ * @brief Extract text and metadata using libarchive.
+ */
+/* Copyright (C) 2020 Parth Kapadia
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include "handler.h"
+
+#include <extractor.h>
+
+#include <cstring>
+
+using namespace std;
+
+string title, keywords, author, pages;
+
+/*
+* cls: passed as last parameter from EXTRACTOR_extract
+* plugin_name:	name of the plugin
+* type: mime-type of file according to libextractor
+* format: format information about data
+* data_mime_type: mimetype of data according to libextractor
+* data: actual meta-data found
+* data_len: number of bytes in data
+*/
+static int
+process_metadata(void* cls,
+		 const char* plugin_name,
+		 enum EXTRACTOR_MetaType type,
+		 enum EXTRACTOR_MetaFormat format,
+		 const char* data_mime_type,
+		 const char* data,
+		 size_t data_len)
+{
+    switch (format) {
+	case EXTRACTOR_METAFORMAT_UTF8:
+	    break;
+	default:
+	    // specific encoding unknown
+	    // EXTRACTOR_METAFORMAT_UNKNOWN
+	    // EXTRACTOR_METAFORMAT_BINARY
+	    // EXTRACTOR_METAFORMAT_C_STRING
+	    return 0;
+    }
+
+    switch (type) {
+	case EXTRACTOR_METATYPE_TITLE:
+	case EXTRACTOR_METATYPE_BOOK_TITLE:
+	case EXTRACTOR_METATYPE_SUBJECT:
+	    title = data;
+	    break;
+
+	case EXTRACTOR_METATYPE_PAGE_COUNT:
+	    pages = data;
+	    break;
+
+	case EXTRACTOR_METATYPE_AUTHOR_NAME:
+	case EXTRACTOR_METATYPE_CREATOR:
+	    author = data;
+	    break;
+
+	default:
+	    keywords += data;
+	    keywords += " ";
+    }
+    return 0;
+}
+
+bool
+extract(const string& filename,
+	const string& mimetype,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    try {
+	// Add all default plugins
+	struct EXTRACTOR_PluginList* plugins
+	 = EXTRACTOR_plugin_add_defaults(EXTRACTOR_OPTION_DEFAULT_POLICY);
+	EXTRACTOR_extract(plugins, filename.c_str(),
+			  NULL, 0,
+			  &process_metadata, NULL);
+	EXTRACTOR_plugin_remove_all(plugins);
+
+	// If plugin not found/ File format not recognised/ corrupt file
+	// Returns null and not error.
+	title = ::title;
+	author = ::author;
+	keywords = ::keywords;
+	pages = ::pages;
+
+	::keywords.resize(0);
+	::title.resize(0);
+	::author.resize(0);
+	::pages.resize(0);
```

Comment:
A better option than resizing the string to zero would be calling the string method [clear](http://www.cplusplus.com/reference/string/string/clear/).

Reply:
Thanks for pointing this out. I was using clear(), I chose to use resize(0) because I read that clear (in case of vector) will make the vector empty but space will still be reserved, So I used resize instead of clear. I am not sure how does it work in case of strings. I will research more about this.
For the handler - I have used a struct as suggested by Olly.

Reply:
https://en.cppreference.com/w/cpp/string/basic_string/clear says:

> Unlike for std::vector::clear, the C++ standard does not explicitly require that capacity is unchanged by this function, but existing implementations do not change capacity. This means that they do not release the allocated memory (see also shrink_to_fit). 

(I'd definitely recommends cppreference.com as a resource - I think there's an offline version too.)

But if you're reusing the strings over and over you probably want the reserved space to be kept.  It's not so helpful if you process one enormous file but mostly small ones, but it's also not helpful to be constantly releasing and reallocating space.